cli/flox/src/commands/gc.rs

```diff
@@ -1,23 +1,291 @@
+    // The `--debug` is intentional here. You don't get any output without it.
+    cmd.args(["store", "gc", "--debug"]);
+    cmd.stderr(Stdio::piped());
+    let mut proc = cmd.spawn().context("Failed to start GC process")?;
+    let stderr = proc
+        .stderr
+        .take()
+        .ok_or(anyhow!("Failed to get stderr from GC process"))?;
+    debug!("spawned gc process");
+
+    let mut gc_progress = GcProgress::new();
+    let reader = BufReader::new(stderr);
+    let (sender, receiver) = std::sync::mpsc::channel::<GcProgress>();
+
+    // It's necessary to keep track of the parent span (the one from the `instrument`
+    // macro) so that we can associate the spans in the spinner thread with it.
+    // Otherwise, you get a spooky spinner message.
+    let parent_span = Span::current();
+
+    // Spawn separate threads for managing the spinner and reading the output
+    // of the `nix store gc` command.
+    std::thread::scope(move |s| {
+        // This thread is responsible for updating the spinner. It's more manageable
+        // to have this as a separate thread with a small inner loop instead of
+        // inlining it into the GC log reader thread.
+        let spinner_thread: std::thread::ScopedJoinHandle<
+            '_,
+            std::result::Result<(), anyhow::Error>,
+        > = s.spawn(move || {
+            let mut inner_span = Some(
+                info_span!(parent: &parent_span, "gc_progress", progress = "Initializing")
+                    .entered(),
+            );
+            let mut done_msg = String::new();
+            // We'll get an error once the sender hangs up.
+            while let Ok(new_progress) = receiver.recv() {
+                if let GcProgress::Done(ref msg) = new_progress {
+                    done_msg = msg.clone();
+                }
+                // This dance is to make sure that the span (and hence the spinner)
+                // stays open until the _start of the next loop body_,
+                // rather than closing at the end of the current loop body
+                // (which is immediately, basically).
+                if let Some(guard) = inner_span.take() {
+                    drop(guard);
+                }
+                inner_span = Some(
+                    info_span!(
+                        parent: &parent_span,
+                        "gc_progress",
+                        progress = new_progress.msg().unwrap_or("Finishing".to_string())
+                    )
+                    .entered(),
+                );
+                debug!(
+                    state = new_progress.state_name(),
+                    "updated spinner with new state"
+                );
+            }
+            debug!("gc log sender hung up, exiting spinner thread");
+
+            // Print a message indicating how many store paths were deleted
+            // and how much disk space was reclaimed.
+            if !done_msg.is_empty() {
```

Comment:
This will probably end up as an `Option`, I was just being lazy when I threw it together.